File: python/simple_tcp_echo.py
"""Module with simple TCP server and client.
"""
import argparse
import socket
import sys

ENCODING = 'utf-8'

def run_server(host: str, port: int):
	with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
		s.bind((host, port))
		s.listen()

		while True:
			print(f'Waiting for a new connection on {host}:{port}')
			c, a = s.accept()

			print(f'Accepted connection from {a}')

			with c:
				msg = c.recv(1024)
				print(f'Received message: ' +
					msg.decode(ENCODING))
				
				# shout back
				c.sendall(msg.decode(ENCODING).upper() \
					.encode(ENCODING))

def run_client(host: str, port: int, msg: str):
	with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
		print(f'Connecting to server {host}:{port}')
		s.connect((host, port))
		print(f'Sending message: {msg}')
		s.sendall(msg[0:1024].encode(ENCODING))
		reply = s.recv(1024).decode(ENCODING)
		print(f'Received reply: {reply}')

if __name__ == '__main__':
	parser = argparse.ArgumentParser()
	parser.add_argument('--server', '-s', dest='server',
		action='store_true', help='Run TCP server')
	parser.add_argument('--client', '-c', dest='client',
		action='store_true', help='Run TCP client')
	
	parser.add_argument('--host', dest='host', required=True,
		type=str, help='Connect to host')
	parser.add_argument('--port', dest='port', required=True,
		type=int, help='Connect to port')
	
	parser.add_argument('--message', '-m', dest='message', required=False,
		type=str, default='Some message', help='Message to echo')


	args = parser.parse_args()

	# Cannot be both server and client, and cannot be neither
	if args.server == args.client:
		print('Cannot be both or neither a server and client')
		sys.exit(1)

	if args.server:
		run_server(args.host, args.port)

	if args.client:
		run_client(args.host, args.port, args.message)

File: c/simple_tcp_echo.c
#include <stdio.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>
#include <unistd.h>

#include "options.h"

struct options options;

void run_server();
void run_client();
void upper(char *s);


int main(int argc, char **argv)
{
	printf("Simple TCP echo start\n");

	options.tcp_udp = TCP;

	int stat = parse_arguments(&options, argc, argv, 1);

	if (stat)
		goto end;

	print_options(&options);

	if (options.client_server == SERVER)
		run_server();
	else
		run_client();

end:
	clear_options(&options);

	printf("Simple TCP echo stop\n");

}

void run_server()
{
	printf("Running server\n");
	int socket_fd = socket(AF_INET, SOCK_STREAM, 0);
	printf("Allocated server fd: %d\n", socket_fd);

	struct sockaddr_in server_address;

	server_address.sin_family = AF_INET;
	server_address.sin_addr.s_addr = inet_addr(options.host);
	server_address.sin_port = htons(options.port);

	int stat;

	if ((stat = bind(socket_fd, (struct sockaddr*)&server_address,
			sizeof(server_address))) != 0) {
		printf("Failed to create server socket, return val %d\n", stat);
		return;
	}

	if (listen(socket_fd, 5) != 0) {
		printf("Failed to listen\n");
		return;
	}

	while(1) {
		printf("Waiting for client connection\n");

		struct sockaddr_in client;
		unsigned int client_size = sizeof(client);
		int client_fd = accept(socket_fd,
			(struct sockaddr*)&client, &client_size);

		if (client_fd < 0) {
			printf("Failed to accept client connection\n");
			return;
		}

		printf("Accepted client connect from \"%s:%d\"\n",
			inet_ntoa(client.sin_addr), client.sin_port);

		char msg[1025];
		memset(msg, 0, 1025);
		recv(client_fd, msg, 1024, 0);

		printf("Client message: \"%s\"\n", msg);

		upper(msg);

		send(client_fd, msg, strlen(msg), 0);
		close(client_fd);
	}

}

void run_client()
{
	printf("Running client\n");

	int socket_fd = socket(AF_INET, SOCK_STREAM, 0);

	if (socket_fd < 0) {
		printf("Created socket with file descriptor %d\n", socket_fd);
		return;
	}

	struct sockaddr_in server_addr = {
		.sin_family = AF_INET,
		.sin_addr.s_addr = inet_addr(options.host),
		.sin_port = htons(options.port)
	};

	int stat = connect(socket_fd,
		(struct sockaddr*)&server_addr, sizeof(server_addr));

	if (stat != 0) {
		printf("Failed to connect to server\n");
		return;
	}

	printf("Sending message to server: %s\n", options.message);

	send(socket_fd, options.message, strlen(options.message), 0);

	char reply[1025];

	// Clear old junk
	memset(reply, 0, 1025);

	recv(socket_fd, reply, strlen(options.message), 0);

	printf("Reply message from server: %s\n", reply);

	close(socket_fd);
}

void upper(char *s)
{
	int len = strlen(s);
	int i;
	for (i = 0; i < len; i++) {
		char upper = toupper(s[i]);
		s[i] = upper;
	}

}
